### DISCUSSION TITLE
[BE201] express session 實作登入功能失敗，req.session.isLogin 讀取為 undefined

### DISCUSSION THREAD

**Discussion by lucktanya33** [Lidemy/forum#191] [後端相關問題] [answered]
# 問題 express session 實作登入功能失敗，req.session.isLogin 讀取為 undefined 

## 問題畫面
輸入正確密碼並成功跳轉，但畫面渲染登入失敗
![](https://i.imgur.com/ddyXqt9.gif)

## 問題敘述
在跟著 BE201 負責管理 Session 的 Session middleware 影片中實作登入功能遇到問題如下：  
在 `index.js` 有設定 req.session.isLogin 但在 `controllers` 那邊拿到 req.session.isLogin 會是 undefined，變成密碼正確但在`views` 裡面渲染出來是登入失敗的狀態。  

###  推測
在 `index.js` 和 `controllers` 裡面分別有把 req.session.isLogin 印出來前者是 // true 後者是 // undefined ，這邊推測所以 `views` 最後拿到 undefined 才會變成登入失敗。

不太曉得為何 req.session.isLogin 的值在 index.js 裡是 // true 傳到 controllers 裡面就變成 // undefined 自己目前推測這是目前造成狀況的主要問題

### 負責 server 的 `index.js`

``` javascript
const express = require('express')
const bodyParser = require('body-parser')
const session = require('express-session')
const app = express() 
const db = require('./db')
const port = 5001

// 使用 middleware: session
app.use(session({
  secret: 'keyboard cat',
  resave: false,
  saveUninitialized: true,
  cookie: { secure: true }
}))

const todoController = require('./controllers/control_todo')
app.set('view engine', 'ejs')

// 使用 middleware: body-parser
app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.post('/test', todoController.newTodo)
app.get('/test', todoController.getAll)
app.get('/test/:id', todoController.get)
app.get('/', todoController.addTodo) // 根目錄放addTodo畫面
app.get('/login', (req, res) => {
  res.render('login')
})
app.post('/login', (req, res) => {
  if (req.body.password === 'abc') {
    req.session.isLogin = true
    console.log('index', req.session.isLogin); // true
    res.redirect('/')
  } else {
    res.redirect('/login')
  }
})
app.get('/logout', (req, res) => {
  req.session.isLogin = false
  res.redirect('/')
})
app.listen(port, () => { 
  db.connect()
  console.log(`Example app listening at http://localhost:${port}`)
})
```
### `controllers`
![](https://i.imgur.com/OXxwBlr.png)

### `views`

```javascript=
<h1>Add todo </h1>

<% if(isLogin) { %>
  你已經登入 <a href ='/logout'>logout</a>
<% } else { %>
  你還沒登入
<% } %>
```

## 嘗試過的解決方案
[express-session使用时 req.session undefined和session.id输出问题](https://blog.csdn.net/weixin_49985592/article/details/109518352)
[req.session is undefined using express-session](https://stackoverflow.com/questions/39796228/req-session-is-undefined-using-express-session)
有查到很多人在討論 req.session是 undefined 的解法是要把 session 的 app use 放在路由的上面，這邊也有做調整但還是一樣的結果

不知道有沒有人在這邊實作的時候是否有遇到類似的問題，或是可能可以往哪個方向去除錯會比較好 
<(_ _*)> 
謝謝看完的大家~~

**Comment by aszx87410** [comment, marked as answer]
把這個 `cookie: { secure: true }` 改成 `cookie: { secure: false }`，因為你 localhost 我猜應該不是 https，所以你設定 secure 的話 cookie 會寫不進去，至於要看有沒有寫進去，可以用 devtool 看 cookie 是不是有成功被傳遞

**Reply by lucktanya33** [reply]
@aszx87410 感謝老師。問題的確就出在這邊，這邊把 secure: true 改成 false ，req.session 馬上就可以寫入東西，成功把狀態傳到 controllers 的檔案裡面了
（這次心得-->debug 找不出原因，果然不一定單純是程式碼的問題，還是要往網路跟瀏覽器的方向去想啊 ><
